react/components/atoms/switcher/switcher.js

```diff
@@ -0,0 +1,137 @@
+import PropTypes from "prop-types";
+import { IdentifiableMixin, baseStyles } from "../../../util";
+import { Touchable } from "../touchable";
+import { mix } from "yonius";
+
+export class Switcher extends mix(PureComponent).with(IdentifiableMixin) {
+    static get propTypes() {
+        return {
+            checked: PropTypes.bool,
+            disabled: PropTypes.bool,
+            checkedText: PropTypes.string,
+            unCheckedText: PropTypes.string,
+            variant: PropTypes.string,
+            animationDuration: PropTypes.number,
+            onValueUpdate: PropTypes.func
+        };
+    }
+
+    static get defaultProps() {
+        return {
+            checked: false,
+            disabled: false,
+            variant: "colored",
+            checkedText: "checked",
+            unCheckedText: "unchecked",
+            animationDuration: 200,
+            onValueUpdate: () => {}
+        };
+    }
+
+    constructor(props) {
+        super(props);
+        this.state = {
+            isChecked: this.props.checked,
+            marginLeft: this.props.checked ? new Animated.Value(20) : new Animated.Value(0),
+            backgroundColor: this.props.checked ? new Animated.Value(150) : new Animated.Value(0)
+        };
+    }
+
+    _getText() {
+        return this.state.isChecked ? this.props.checkedText : this.props.unCheckedText;
+    }
+
+    _style = () => {
+        const isColored = this.props.variant === "colored";
+        return [
+            styles.switcher,
+            {
+                backgroundColor: this.state.backgroundColor.interpolate({
+                    inputRange: [0, 150],
+                    outputRange: ["#cccccc", isColored ? "#507BF8" : "#1d1d1d"]
+                }),
+                borderColor: this.state.backgroundColor.interpolate({
+                    inputRange: [0, 150],
+                    outputRange: ["#cccccc", isColored ? "#507BF8" : "#1d1d1d"]
+                }),
+                opacity: this.props.disabled ? 0.3 : 1
+            }
+        ];
+    };
+
+    _buttonStyle = () => {
+        return [styles.button, { marginLeft: this.state.marginLeft }];
+    };
+
+    onPress = () => {
+        const marginLeftValue = this.state.isChecked ? 0 : 20;
+        const backgroundColor = this.state.isChecked ? 0 : 150;
+        Animated.parallel([
+            Animated.timing(this.state.marginLeft, {
+                toValue: marginLeftValue,
+                duration: this.props.animationDuration,
+                useNativeDriver: false
+            }),
+            Animated.timing(this.state.backgroundColor, {
+                toValue: backgroundColor,
+                duration: this.props.animationDuration,
+                useNativeDriver: false
+            })
+        ]).start(() => {
+            this.setState({ isChecked: !this.state.isChecked }, () =>
+                this.props.onValueUpdate(this.state.isChecked)
+            );

```

Comment:
I believe it is a good pratice to use the prevState provided by the setState, something like:
```
this.setState(prevState => ({ isChecked: !prevState.isChecked}), () => ....)
 ```